Multipart form data is interpreted as a file although it's not a file

### Describe the bug

If multipart form data is not a file and specified with other MIME-Type than `text/plain`, quarkus tries to parse form data as a file, which leads to a NullPointerException.

Was introduced in quarkus 3.7.3 with this change: https://github.com/quarkusio/quarkus/pull/38714

### Expected behavior

Multipart form data is parsed with correct content type.

### Actual behavior

Multipart form data other than `text/plain` is parsed as File.

Request:

```
Hypertext Transfer Protocol
MIME Multipart Media Encapsulation, Type: multipart/form-data, Boundary: "5vI261AD6aFM_WL55jbNMpcLal_Lzt2uTnRW"
    [Type: multipart/form-data]
    First boundary: --5vI261AD6aFM_WL55jbNMpcLal_Lzt2uTnRW\r\n
    Encapsulated multipart part:  (application/json)
        Content-Disposition: form-data; name="firstParam"\r\n
        Content-Type: application/json; charset=US-ASCII\r\n
        Content-Transfer-Encoding: 8bit\r\n\r\n
        JavaScript Object Notation: application/json
    Last boundary: \r\n--5vI261AD6aFM_WL55jbNMpcLal_Lzt2uTnRW--\r\n
```

Exception:

```
2024-02-15 15:58:14,745 DEBUG [WebApplicationException] (executor-thread-1) Restarting handler chain for exception exception: jakarta.ws.rs.WebApplicationException: HTTP 400 Bad Request
	at org.jboss.resteasy.reactive.server.handlers.ParameterHandler.handle(ParameterHandler.java:66)
	at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:137)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:582)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.NullPointerException
	at java.base/java.util.Objects.requireNonNull(Objects.java:233)
	at java.base/java.nio.file.Files.readString(Files.java:3343)
	at org.jboss.resteasy.reactive.server.core.multipart.MultipartSupport.getString(MultipartSupport.java:204)
	at org.jboss.resteasy.reactive.server.core.parameters.MultipartFormParamExtractor.extractParameter(MultipartFormParamExtractor.java:46)
	at org.jboss.resteasy.reactive.server.handlers.ParameterHandler.handle(ParameterHandler.java:45)
	... 9 more
```

### How to Reproduce?

I've made a reproducer: https://github.com/chutzemischt/quarkus-multipart-reproducer

```java
@Path("/api")
public class MyTestController {

    @Path("test")
    @POST
    public Response testMultipart(@FormParam("firstParam") final String firstParam, @FormParam("secondParam") final String secondParam) {
        return Response.ok().build();
    }
}
```

```java
@QuarkusTest
class MyTestControllerTest {

    @Test
    void testMultipart() {
        given().multiPart("firstParam", "{\"id\":\"myId\",\"name\":\"myName\"}", "application/json")
                .when()
                .post("/api/test")
                .then()
                .statusCode(200);
    }
}
```

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

21.0.1

### Quarkus version or git rev

3.7.3

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Gradle 8.5

### Additional information

_No response_

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/multipart/MultipartTextWithoutFilenameTest.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/multipart/MultipartTextWithoutFilenameTest.java
@@ -1,0 +1,52 @@
+package io.quarkus.resteasy.reactive.server.test.multipart;
+
+import static io.restassured.RestAssured.given;
+
+import java.io.IOException;
+import java.util.function.Supplier;
+
+import jakarta.ws.rs.FormParam;
+import jakarta.ws.rs.POST;
+import jakarta.ws.rs.Path;
+
+import org.jboss.resteasy.reactive.RestResponse;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+public class MultipartTextWithoutFilenameTest {
+
+    @RegisterExtension
+    static QuarkusUnitTest test = new QuarkusUnitTest()
+            .setArchiveProducer(new Supplier<>() {
+                @Override
+                public JavaArchive get() {
+                    return ShrinkWrap.create(JavaArchive.class)
+                            .addClasses(Resource.class);
+                }
+            });
+
+    @Test
+    public void test() throws IOException {
+        given()
+                .contentType("multipart/form-data")
+                .multiPart("firstParam", "{\"id\":\"myId\",\"name\":\"myName\"}", "application/json")
+                .when()
+                .post("/test")
+                .then()
+                .statusCode(200);
+    }
+
+    @Path("/test")
+    public static class Resource {
+
+        @POST
+        public RestResponse<Void> testMultipart(@FormParam("firstParam") final String firstParam,
+                @FormParam("secondParam") final String secondParam) {
+            return RestResponse.ok();
+        }
+    }
+}

--- a/independent-projects/resteasy-reactive/common/runtime/src/main/java/org/jboss/resteasy/reactive/common/util/MediaTypeHelper.java
+++ b/independent-projects/resteasy-reactive/common/runtime/src/main/java/org/jboss/resteasy/reactive/common/util/MediaTypeHelper.java
@@ -267,10 +267,12 @@ public class MediaTypeHelper {
         return true;
     }
 
+    // TODO: does this need to be more complex?
     public static boolean isTextLike(MediaType mediaType) {
-        return "text".equalsIgnoreCase(mediaType.getType())
-                || ("application".equalsIgnoreCase(mediaType.getType())
-                        && mediaType.getSubtype().toLowerCase().startsWith("xml"));
+        String type = mediaType.getType();
+        String subtype = mediaType.getSubtype();
+        return (type.equals("application") && (subtype.contains("json") || subtype.contains("xml") || subtype.contains("yaml")))
+                || type.equals("text");
     }
 
     public static boolean isUnsupportedWildcardSubtype(MediaType mediaType) {

--- a/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/EncodedMediaType.java
+++ b/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/EncodedMediaType.java
@@ -5,6 +5,7 @@ import java.util.Objects;
 
 import jakarta.ws.rs.core.MediaType;
 
+import org.jboss.resteasy.reactive.common.util.MediaTypeHelper;
 import org.jboss.resteasy.reactive.server.spi.ContentType;
 
 /**
@@ -21,7 +22,7 @@ public class EncodedMediaType implements ContentType {
         MediaType effectiveMediaType = mediaType;
         String effectiveCharset;
         String originalCharset = mediaType.getParameters().get("charset");
-        if (isStringMediaType(mediaType)) {
+        if (MediaTypeHelper.isTextLike(mediaType)) {
             effectiveCharset = originalCharset;
             if (effectiveCharset == null) {
                 effectiveCharset = StandardCharsets.UTF_8.name();
@@ -38,12 +39,6 @@ public class EncodedMediaType implements ContentType {
     }
 
     // TODO: does this need to be more complex?
-    private boolean isStringMediaType(MediaType mediaType) {
-        String type = mediaType.getType();
-        String subtype = mediaType.getSubtype();
-        return (type.equals("application") && (subtype.contains("json") || subtype.contains("xml") || subtype.contains("yaml")))
-                || type.equals("text");
-    }
 
     @Override
     public String toString() {

--- a/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/multipart/MultiPartParserDefinition.java
+++ b/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/multipart/MultiPartParserDefinition.java
@@ -29,6 +29,7 @@ import jakarta.ws.rs.core.Response;
 import org.jboss.logging.Logger;
 import org.jboss.resteasy.reactive.common.headers.HeaderUtil;
 import org.jboss.resteasy.reactive.common.util.CaseInsensitiveMap;
+import org.jboss.resteasy.reactive.common.util.MediaTypeHelper;
 import org.jboss.resteasy.reactive.server.core.ResteasyReactiveRequestContext;
 import org.jboss.resteasy.reactive.server.spi.ServerHttpRequest;
 
@@ -366,7 +367,7 @@ public class MultiPartParserDefinition implements FormParserFactory.ParserDefini
             if (contentType == null || contentType.isEmpty()) { // https://www.rfc-editor.org/rfc/rfc7578.html#section-4.4 says the default content-type if missing is text/plain
                 return true;
             }
-            return MediaType.TEXT_PLAIN_TYPE.isCompatible(MediaType.valueOf(contentType));
+            return MediaTypeHelper.isTextLike(MediaType.valueOf(contentType));
         }
 
         public List<Path> getCreatedFiles() {

